Generic SFC broken from 1.8.10, 1.8.11

I had a generic component that was working, but it isn't resolving correctly in version 1.8.10 and 1.8.11, but if I downgrade to 1.8.8, it works fine.

In this example in index.js, `item` is just resolving to `unknown`, which results in a lot of errors.

### index.vue
```JS
<CarouselSelect
  v-slot="{ item }"
  v-model="componentState.selected"
  :items="componentState.items"
  :get-key="(item) => item.id"
  class="mb-3"
>
  {{item.title}}
</CarouselSelect>
```

### CarouselSelect.vue
```JS
<script setup lang="ts" generic="T, TValue extends undefined|string|number|symbol">
const $props = defineProps<{
  items: Array<T>,
  modelValue: TValue,
  getKey: (item: T) => TValue,
}>();

defineSlots<{
  default: { item: T },
}>();
//...
</script>
```

> The version range is incorrect. Every version I tested reports this error.

Can't confirm that.

Reproduction: https://stackblitz.com/edit/nuxt-starter-hqnmsd?file=components%2FGenericComponent.vue,app.vue

```
❯ npm install vue-tsc@1.8.8

changed 3 packages in 5s

91 packages are looking for funding
  run `npm fund` for details

~/projects/nuxt-starter-hqnmsd 5s
❯ npm run typecheck

> typecheck
> nuxt prepare && nuxt typecheck

ℹ [nuxt] compiled plugins/server.mjs in 1245.01ms                                  21:21:32
ℹ [nuxt] compiled plugins/client.mjs in 1308.49ms                                  21:21:32
✔ Types generated in .nuxt                                                         21:21:33
ℹ [nuxt] compiled plugins/server.mjs in 2694.39ms                                  21:21:42
ℹ [nuxt] compiled plugins/client.mjs in 2946.65ms                                  21:21:42

~/projects/nuxt-starter-hqnmsd 24s
```

vs.

```
❯ npm install vue-tsc@1.8.11

changed 3 packages in 5s

91 packages are looking for funding
  run `npm fund` for details

~/projects/nuxt-starter-hqnmsd 5s
❯ npm run typecheck

> typecheck
> nuxt prepare && nuxt typecheck

ℹ [nuxt] compiled plugins/server.mjs in 1382.58ms                                  21:22:21
ℹ [nuxt] compiled plugins/client.mjs in 1473.76ms                                  21:22:21
✔ Types generated in .nuxt                                                         21:22:21
ℹ [nuxt] compiled plugins/server.mjs in 4979.01ms                                  21:22:36
ℹ [nuxt] compiled plugins/client.mjs in 5085.85ms                                  21:22:36
app.vue:5:25 - error TS18046: 'el' is of type 'unknown'.

5       :orderBy="(el) => el.toUpperCase()"
                          ~~

app.vue:8:12 - error TS18046: 'item' is of type 'unknown'.

8         {{ item.toUpperCase() }}
             ~~~~


Found 2 errors in the same file, starting at: app.vue:5


 ERROR  Command failed with exit code 2: vue-tsc --noEmit                           21:22:50

  at makeError (node_modules/nuxi/dist/chunks/index3.mjs:645:11)
  at handlePromise (node_modules/nuxi/dist/chunks/index3.mjs:1841:26)
  at async Object.run (node_modules/nuxi/dist/chunks/typecheck.mjs:94:7)
  at async runCommand$1 (node_modules/nuxi/dist/shared/nuxi.89381a23.mjs:1674:7)
  at async runCommand$1 (node_modules/nuxi/dist/shared/nuxi.89381a23.mjs:1665:11)
  at async runMain$1 (node_modules/nuxi/dist/shared/nuxi.89381a23.mjs:1797:7) 



 ERROR  Command failed with exit code 2: vue-tsc --noEmit                           21:22:50
```

--- a/packages/vue-language-core/src/generators/template.ts
+++ b/packages/vue-language-core/src/generators/template.ts
@@ -749,7 +749,7 @@ export function generate(
 
 			codes.push(
 				`type ${var_componentPropsInstance}_props = NonNullable<NonNullable<typeof ${var_componentPropsInstance}.__ctx>['props']>;\n`,
-				`type ${var_componentInstance}_propsType = __VLS_MappedOmit<Parameters<typeof ${var_functionalComponent}>[0], (__VLS_IsAny<${var_componentPropsInstance}_props> extends true ? never : keyof ${var_componentPropsInstance}_props) | keyof JSX.ElementChildrenAttribute> & __VLS_MappedOmit<__VLS_PickNotAny<${var_componentPropsInstance}_props, {}>, keyof JSX.ElementChildrenAttribute>;\n`,
+				`type ${var_componentInstance}_propsType = __VLS_MappedOmit<Parameters<typeof ${var_functionalComponent}>[0], (__VLS_IsAny<${var_componentPropsInstance}_props> extends true ? never : keyof ${var_componentPropsInstance}_props) | keyof JSX.ElementChildrenAttribute> & __VLS_MappedOmit<__VLS_PickNotAny<${var_componentPropsInstance}_props, {}>, keyof JSX.ElementChildrenAttribute> & Record<string, unknown>;\n`,
 			);
 		}
 
@@ -759,10 +759,11 @@ export function generate(
 			tagOffsets.length ? ['', 'template', tagOffsets[0], capabilitiesPresets.diagnosticOnly]
 				: dynamicTagExp ? ['', 'template', startTagOffset, capabilitiesPresets.diagnosticOnly]
 					: '',
+			vueCompilerOptions.strictTemplates ? '' : `__VLS_cast<__VLS_Prettify<${var_componentInstance}_propsType>>(`,
 			'{ ',
 			...createPropsCode(node, props, 'normal', propsFailedExps),
 			`}`,
-			vueCompilerOptions.strictTemplates ? '' : ` as ${var_componentInstance}_propsType`,
+			vueCompilerOptions.strictTemplates ? '' : `)`,
 			// diagnostic end
 			tagOffsets.length ? ['', 'template', tagOffsets[0] + tag.length, capabilitiesPresets.diagnosticOnly]
 				: dynamicTagExp ? ['', 'template', startTagOffset + tag.length, capabilitiesPresets.diagnosticOnly]

--- a/packages/vue-language-core/src/generators/template.ts
+++ b/packages/vue-language-core/src/generators/template.ts
@@ -748,8 +748,8 @@ export function generate(
 			);
 
 			codes.push(
-				`type ${var_componentPropsInstance}_props = NonNullable<typeof ${var_componentPropsInstance}.__ctx>['props'];\n`,
-				`type ${var_componentInstance}_propsType = Omit<Parameters<typeof ${var_functionalComponent}>[0], (__VLS_IsAny<${var_componentPropsInstance}_props> extends true ? never : keyof ${var_componentPropsInstance}_props) | JSX.ElementChildrenAttribute> & __VLS_PickNotAny<${var_componentPropsInstance}_props, {}>;\n`,
+				`type ${var_componentPropsInstance}_props = NonNullable<NonNullable<typeof ${var_componentPropsInstance}.__ctx>['props']>;\n`,
+				`type ${var_componentInstance}_propsType = __VLS_MappedOmit<Parameters<typeof ${var_functionalComponent}>[0], (__VLS_IsAny<${var_componentPropsInstance}_props> extends true ? never : keyof ${var_componentPropsInstance}_props) | keyof JSX.ElementChildrenAttribute> & __VLS_MappedOmit<__VLS_PickNotAny<${var_componentPropsInstance}_props, {}>, keyof JSX.ElementChildrenAttribute>;\n`,
 			);
 		}
 

--- a/packages/vue-language-core/src/generators/template.ts
+++ b/packages/vue-language-core/src/generators/template.ts
@@ -748,8 +748,8 @@ export function generate(
 			);
 
 			codes.push(
-				`type ${var_componentPropsInstance}_props = NonNullable<NonNullable<typeof ${var_componentPropsInstance}.__ctx>['props']>;\n`,
-				`type ${var_componentInstance}_propsType = __VLS_MappedOmit<Parameters<typeof ${var_functionalComponent}>[0], (__VLS_IsAny<${var_componentPropsInstance}_props> extends true ? never : keyof ${var_componentPropsInstance}_props) | keyof JSX.ElementChildrenAttribute> & __VLS_MappedOmit<__VLS_PickNotAny<${var_componentPropsInstance}_props, {}>, keyof JSX.ElementChildrenAttribute> & Record<string, unknown>;\n`,
+				`type ${var_componentPropsInstance}_props = __VLS_GetComponentProps<typeof ${var_componentPropsInstance}>;\n`,
+				`type ${var_componentInstance}_propsType = __VLS_GetComponentPropsType<typeof ${var_functionalComponent}, ${var_componentPropsInstance}_props>;\n`,
 			);
 		}
 

--- a/packages/vue-language-core/src/generators/template.ts
+++ b/packages/vue-language-core/src/generators/template.ts
@@ -1176,7 +1176,6 @@ export function generate(
 			};
 		}
 
-
 		codes.push(`...{ `);
 		for (const prop of props) {
 			if (
@@ -1790,8 +1789,14 @@ export function generate(
 
 		codes.push('// @ts-ignore\n'); // #2304
 		codes.push('[');
+
+		const usedVars = new Set<string>();
+
 		for (const _vars of tempVars) {
 			for (const v of _vars) {
+				if (usedVars.has(v.text)) continue;
+
+				usedVars.add(v.text);
 				codes.push([v.text, 'template', v.offset, { completion: { additional: true } }]);
 				codes.push(',');
 			}

--- a/packages/vue-language-core/src/generators/template.ts
+++ b/packages/vue-language-core/src/generators/template.ts
@@ -749,7 +749,7 @@ export function generate(
 
 			codes.push(
 				`type ${var_componentPropsInstance}_props = NonNullable<typeof ${var_componentPropsInstance}.__ctx>['props'];\n`,
-				`type ${var_componentInstance}_propsType = Omit<Parameters<typeof ${var_functionalComponent}>[0], __VLS_IsAny<${var_componentPropsInstance}_props> extends true ? never : keyof ${var_componentPropsInstance}_props> & __VLS_PickNotAny<${var_componentPropsInstance}_props, {}>;\n`,
+				`type ${var_componentInstance}_propsType = Omit<Parameters<typeof ${var_functionalComponent}>[0], (__VLS_IsAny<${var_componentPropsInstance}_props> extends true ? never : keyof ${var_componentPropsInstance}_props) | JSX.ElementChildrenAttribute> & __VLS_PickNotAny<${var_componentPropsInstance}_props, {}>;\n`,
 			);
 		}
 

--- a/packages/vue-language-core/src/generators/template.ts
+++ b/packages/vue-language-core/src/generators/template.ts
@@ -622,6 +622,7 @@ export function generate(
 		const isNamespacedTag = tag.indexOf('.') >= 0;
 		const var_originalComponent = `__VLS_${elementIndex++}`;
 		const var_functionalComponent = `__VLS_${elementIndex++}`;
+		const var_componentPropsInstance = `__VLS_${elementIndex++}`;
 		const var_componentInstance = `__VLS_${elementIndex++}`;
 
 		let dynamicTagExp: CompilerDOM.ExpressionNode | undefined;
@@ -737,6 +738,21 @@ export function generate(
 			}
 		}
 
+		if (!vueCompilerOptions.strictTemplates) {
+			codes.push(
+				`const ${var_componentPropsInstance} = ${var_functionalComponent}(`,
+				'{ ',
+				...createPropsCode(node, props, 'noFormat', propsFailedExps).map(code => typeof code === 'string' ? code : code[0]),
+				'}',
+				`, ...__VLS_functionalComponentArgsRest(${var_functionalComponent}));\n`,
+			);
+
+			codes.push(
+				`type ${var_componentPropsInstance}_props = NonNullable<typeof ${var_componentPropsInstance}.__ctx>['props'];\n`,
+				`type ${var_componentInstance}_propsType = Omit<Parameters<typeof ${var_functionalComponent}>[0], __VLS_IsAny<${var_componentPropsInstance}_props> extends true ? never : keyof ${var_componentPropsInstance}_props> & __VLS_PickNotAny<${var_componentPropsInstance}_props, {}>;\n`,
+			);
+		}
+
 		codes.push(
 			`const ${var_componentInstance} = ${var_functionalComponent}(`,
 			// diagnostic start
@@ -744,17 +760,9 @@ export function generate(
 				: dynamicTagExp ? ['', 'template', startTagOffset, capabilitiesPresets.diagnosticOnly]
 					: '',
 			'{ ',
-		);
-		if (!vueCompilerOptions.strictTemplates) {
-			// fix https://github.com/vuejs/language-tools/issues/3318
-			codes.push('...{ ');
-		}
-		codes.push(...createPropsCode(node, props, 'normal', propsFailedExps));
-		if (!vueCompilerOptions.strictTemplates) {
-			codes.push('}, ');
-		}
-		codes.push(
-			'}',
+			...createPropsCode(node, props, 'normal', propsFailedExps),
+			`}`,
+			vueCompilerOptions.strictTemplates ? '' : ` as ${var_componentInstance}_propsType`,
 			// diagnostic end
 			tagOffsets.length ? ['', 'template', tagOffsets[0] + tag.length, capabilitiesPresets.diagnosticOnly]
 				: dynamicTagExp ? ['', 'template', startTagOffset + tag.length, capabilitiesPresets.diagnosticOnly]
@@ -1130,7 +1138,7 @@ export function generate(
 		}
 	}
 
-	function createPropsCode(node: CompilerDOM.ElementNode, props: CompilerDOM.ElementNode['props'], mode: 'normal' | 'extraReferences', propsFailedExps?: CompilerDOM.SimpleExpressionNode[]): Code[] {
+	function createPropsCode(node: CompilerDOM.ElementNode, props: CompilerDOM.ElementNode['props'], mode: 'noFormat' | 'normal' | 'extraReferences', propsFailedExps?: CompilerDOM.SimpleExpressionNode[]): Code[] {
 
 		let styleAttrNum = 0;
 		let classAttrNum = 0;

--- a/packages/vue-language-core/src/generators/template.ts
+++ b/packages/vue-language-core/src/generators/template.ts
@@ -622,7 +622,6 @@ export function generate(
 		const isNamespacedTag = tag.indexOf('.') >= 0;
 		const var_originalComponent = `__VLS_${elementIndex++}`;
 		const var_functionalComponent = `__VLS_${elementIndex++}`;
-		const var_componentPropsInstance = `__VLS_${elementIndex++}`;
 		const var_componentInstance = `__VLS_${elementIndex++}`;
 
 		let dynamicTagExp: CompilerDOM.ExpressionNode | undefined;
@@ -738,35 +737,50 @@ export function generate(
 			}
 		}
 
-		if (!vueCompilerOptions.strictTemplates) {
+		if (vueCompilerOptions.strictTemplates) {
+			// with strictTemplates, generate once for props type-checking + instance type
 			codes.push(
-				`const ${var_componentPropsInstance} = ${var_functionalComponent}(`,
+				`const ${var_componentInstance} = ${var_functionalComponent}(`,
+				// diagnostic start
+				tagOffsets.length ? ['', 'template', tagOffsets[0], capabilitiesPresets.diagnosticOnly]
+					: dynamicTagExp ? ['', 'template', startTagOffset, capabilitiesPresets.diagnosticOnly]
+						: '',
 				'{ ',
-				...createPropsCode(node, props, 'noFormat', propsFailedExps).map(code => typeof code === 'string' ? code : code[0]),
+				...createPropsCode(node, props, 'normal', propsFailedExps),
+				'}',
+				// diagnostic end
+				tagOffsets.length ? ['', 'template', tagOffsets[0] + tag.length, capabilitiesPresets.diagnosticOnly]
+					: dynamicTagExp ? ['', 'template', startTagOffset + tag.length, capabilitiesPresets.diagnosticOnly]
+						: '',
+				`, ...__VLS_functionalComponentArgsRest(${var_functionalComponent}));\n`,
+			);
+		}
+		else {
+			// without strictTemplates, this only for instacne type
+			codes.push(
+				`const ${var_componentInstance} = ${var_functionalComponent}(`,
+				'{ ',
+				...createPropsCode(node, props, 'extraReferences'),
 				'}',
 				`, ...__VLS_functionalComponentArgsRest(${var_functionalComponent}));\n`,
 			);
-
-			codes.push(`type ${var_componentInstance}_propsType = __VLS_GetComponentPropsType<typeof ${var_functionalComponent}, __VLS_GetComponentProps<typeof ${var_componentPropsInstance}>>;\n`);
+			// and this for props type-checking
+			codes.push(
+				`({} as (props: __VLS_FunctionalComponentProps<typeof ${var_originalComponent}, typeof ${var_componentInstance}> & Record<string, unknown>) => void)(`,
+				// diagnostic start
+				tagOffsets.length ? ['', 'template', tagOffsets[0], capabilitiesPresets.diagnosticOnly]
+					: dynamicTagExp ? ['', 'template', startTagOffset, capabilitiesPresets.diagnosticOnly]
+						: '',
+				'{ ',
+				...createPropsCode(node, props, 'normal', propsFailedExps),
+				'}',
+				// diagnostic end
+				tagOffsets.length ? ['', 'template', tagOffsets[0] + tag.length, capabilitiesPresets.diagnosticOnly]
+					: dynamicTagExp ? ['', 'template', startTagOffset + tag.length, capabilitiesPresets.diagnosticOnly]
+						: '',
+				`);\n`,
+			);
 		}
-
-		codes.push(
-			`const ${var_componentInstance} = ${var_functionalComponent}(`,
-			// diagnostic start
-			tagOffsets.length ? ['', 'template', tagOffsets[0], capabilitiesPresets.diagnosticOnly]
-				: dynamicTagExp ? ['', 'template', startTagOffset, capabilitiesPresets.diagnosticOnly]
-					: '',
-			vueCompilerOptions.strictTemplates ? '' : `__VLS_cast<__VLS_Prettify<${var_componentInstance}_propsType>>(`,
-			'{ ',
-			...createPropsCode(node, props, 'normal', propsFailedExps),
-			`}`,
-			vueCompilerOptions.strictTemplates ? '' : `)`,
-			// diagnostic end
-			tagOffsets.length ? ['', 'template', tagOffsets[0] + tag.length, capabilitiesPresets.diagnosticOnly]
-				: dynamicTagExp ? ['', 'template', startTagOffset + tag.length, capabilitiesPresets.diagnosticOnly]
-					: '',
-			`, ...__VLS_functionalComponentArgsRest(${var_functionalComponent}));\n`,
-		);
 
 		if (tag !== 'template' && tag !== 'slot') {
 			componentCtxVar = `__VLS_${elementIndex++}`;
@@ -986,7 +1000,7 @@ export function generate(
 				const eventVar = `__VLS_${elementIndex++}`;
 				codes.push(
 					`let ${eventVar} = { '${prop.arg.loc.source}': `,
-					`__VLS_pickEvent(${componentCtxVar}.emit!, '${prop.arg.loc.source}' as const, __VLS_componentProps(${componentVar}, ${componentInstanceVar})`,
+					`__VLS_pickEvent(${componentCtxVar}.emit!, '${prop.arg.loc.source}' as const, {} as __VLS_FunctionalComponentProps<typeof ${componentVar}, typeof ${componentInstanceVar}>`,
 					...createPropertyAccessCode([
 						camelize('on-' + prop.arg.loc.source), // onClickOutside
 						'template',
@@ -1136,7 +1150,7 @@ export function generate(
 		}
 	}
 
-	function createPropsCode(node: CompilerDOM.ElementNode, props: CompilerDOM.ElementNode['props'], mode: 'noFormat' | 'normal' | 'extraReferences', propsFailedExps?: CompilerDOM.SimpleExpressionNode[]): Code[] {
+	function createPropsCode(node: CompilerDOM.ElementNode, props: CompilerDOM.ElementNode['props'], mode: 'normal' | 'extraReferences', propsFailedExps?: CompilerDOM.SimpleExpressionNode[]): Code[] {
 
 		let styleAttrNum = 0;
 		let classAttrNum = 0;
@@ -1786,14 +1800,8 @@ export function generate(
 
 		codes.push('// @ts-ignore\n'); // #2304
 		codes.push('[');
-
-		const usedVars = new Set<string>();
-
 		for (const _vars of tempVars) {
 			for (const v of _vars) {
-				if (usedVars.has(v.text)) continue;
-
-				usedVars.add(v.text);
 				codes.push([v.text, 'template', v.offset, { completion: { additional: true } }]);
 				codes.push(',');
 			}

--- a/packages/vue-language-core/src/generators/template.ts
+++ b/packages/vue-language-core/src/generators/template.ts
@@ -747,10 +747,7 @@ export function generate(
 				`, ...__VLS_functionalComponentArgsRest(${var_functionalComponent}));\n`,
 			);
 
-			codes.push(
-				`type ${var_componentPropsInstance}_props = __VLS_GetComponentProps<typeof ${var_componentPropsInstance}>;\n`,
-				`type ${var_componentInstance}_propsType = __VLS_GetComponentPropsType<typeof ${var_functionalComponent}, ${var_componentPropsInstance}_props>;\n`,
-			);
+			codes.push(`type ${var_componentInstance}_propsType = __VLS_GetComponentPropsType<typeof ${var_functionalComponent}, __VLS_GetComponentProps<typeof ${var_componentPropsInstance}>>;\n`);
 		}
 
 		codes.push(

--- a/packages/vue-language-core/src/utils/globalTypes.ts
+++ b/packages/vue-language-core/src/utils/globalTypes.ts
@@ -97,7 +97,7 @@ declare function __VLS_asFunctionalComponent<T, K = T extends new (...args: any)
 	}) => JSX.Element & { __ctx?: typeof ctx & { props?: typeof props; expose?(exposed: K): void; } }
 	: T extends () => any ? (props: {}, ctx?: any) => ReturnType<T>
 	: T extends (...args: any) => any ? T
-	: (_: T extends import('${vueCompilerOptions.lib}').VNode | import('${vueCompilerOptions.lib}').VNode[] | string ? {}: T${vueCompilerOptions.strictTemplates ? '' : ' & Record<string, unknown>'}, ctx?: any) => { __ctx?: { attrs?: any, expose?: any, slots?: any, emit?: any, props?: T${vueCompilerOptions.strictTemplates ? '' : ' & Record<string, unknown>'} } }; // IntrinsicElement
+	: (_: T extends import('${vueCompilerOptions.lib}').VNode | import('${vueCompilerOptions.lib}').VNode[] | string ? {}: T${vueCompilerOptions.strictTemplates ? '' : ' & Record<string, unknown>'}, ctx?: any) => { __ctx?: { attrs?: any, expose?: any, slots?: any, emit?: any, props?: (T extends object ? T : {})${vueCompilerOptions.strictTemplates ? '' : ' & Record<string, unknown>'} } }; // IntrinsicElement
 declare function __VLS_functionalComponentArgsRest<T extends (...args: any) => any>(t: T): Parameters<T>['length'] extends 2 ? [any] : [];
 declare function __VLS_pickEvent<Emit, K, E>(emit: Emit, emitKey: K, event: E): __VLS_FillingEventArg<
 	__VLS_PickNotAny<

--- a/packages/vue-language-core/src/utils/globalTypes.ts
+++ b/packages/vue-language-core/src/utils/globalTypes.ts
@@ -119,6 +119,8 @@ declare function __VLS_componentProps<T, K>(comp: T, fnReturn: K):
 	: T extends (props: infer P, ...args: any) => any ? NonNullable<P> :
 	{};
 
+type __VLS_GetComponentProps<T> = NonNullable<NonNullable<T['__ctx']>['props']>;
+type __VLS_GetComponentPropsType<C, P> = __VLS_MappedOmit<Parameters<C>[0], (__VLS_IsAny<P> extends true ? never : keyof P) | keyof JSX.ElementChildrenAttribute> & __VLS_MappedOmit<__VLS_PickNotAny<P, {}>, keyof JSX.ElementChildrenAttribute> & Record<string, unknown>;
 declare function __VLS_cast<T>(x: T): T;
 `.trim();
 }

--- a/packages/vue-language-core/src/utils/globalTypes.ts
+++ b/packages/vue-language-core/src/utils/globalTypes.ts
@@ -13,6 +13,8 @@ type __VLS_Element = __VLS_PickNotAny<import('vue/jsx-runtime').JSX.Element, JSX
 type __VLS_IsAny<T> = 0 extends 1 & T ? true : false;
 type __VLS_PickNotAny<A, B> = __VLS_IsAny<A> extends true ? B : A;
 
+type __VLS_MappedOmit<T, K> = { [P in keyof T as P extends K ? never : P]: T[P] };
+
 type __VLS_Prettify<T> = {
 	[K in keyof T]: T[K];
 } & {};

--- a/packages/vue-language-core/src/utils/globalTypes.ts
+++ b/packages/vue-language-core/src/utils/globalTypes.ts
@@ -99,7 +99,7 @@ declare function __VLS_asFunctionalComponent<T, K = T extends new (...args: any)
 	}) => JSX.Element & { __ctx?: typeof ctx & { props?: typeof props; expose?(exposed: K): void; } }
 	: T extends () => any ? (props: {}, ctx?: any) => ReturnType<T>
 	: T extends (...args: any) => any ? T
-	: (_: T extends import('${vueCompilerOptions.lib}').VNode | import('${vueCompilerOptions.lib}').VNode[] | string ? {}: T${vueCompilerOptions.strictTemplates ? '' : ' & Record<string, unknown>'}, ctx?: any) => { __ctx?: { attrs?: any, expose?: any, slots?: any, emit?: any, props?: (T extends object ? T : {})${vueCompilerOptions.strictTemplates ? '' : ' & Record<string, unknown>'} } }; // IntrinsicElement
+	: (_: T extends import('${vueCompilerOptions.lib}').VNode | import('${vueCompilerOptions.lib}').VNode[] | string ? {}: T${vueCompilerOptions.strictTemplates ? '' : ' & Record<string, unknown>'}, ctx?: any) => { __ctx?: { attrs?: any, expose?: any, slots?: any, emit?: any, props?: (T extends import('${vueCompilerOptions.lib}').VNode ? {} : T extends object ? T : {})${vueCompilerOptions.strictTemplates ? '' : ' & Record<string, unknown>'} } }; // IntrinsicElement
 declare function __VLS_functionalComponentArgsRest<T extends (...args: any) => any>(t: T): Parameters<T>['length'] extends 2 ? [any] : [];
 declare function __VLS_pickEvent<Emit, K, E>(emit: Emit, emitKey: K, event: E): __VLS_FillingEventArg<
 	__VLS_PickNotAny<
@@ -108,20 +108,16 @@ declare function __VLS_pickEvent<Emit, K, E>(emit: Emit, emitKey: K, event: E): 
 	>
 > | undefined;
 declare function __VLS_pickFunctionalComponentCtx<T, K>(comp: T, compInstance: K): __VLS_PickNotAny<
-	K extends { __ctx?: infer Ctx } ? Ctx : any,
-	T extends (props: any, ctx: infer Ctx) => any ? Ctx : any
+	'__ctx' extends keyof __VLS_PickNotAny<K, {}> ? K extends { __ctx?: infer Ctx } ? Ctx : never : any
+	, T extends (props: any, ctx: infer Ctx) => any ? Ctx : any
 >;
+type __VLS_FunctionalComponentProps<T, K> =
+	'__ctx' extends keyof __VLS_PickNotAny<K, {}> ? K extends { __ctx?: { props?: infer P } } ? NonNullable<P> : never
+	: T extends (props: infer P, ...args: any) => any ? P :
+	{};
 type __VLS_AsFunctionOrAny<F> = unknown extends F ? any : ((...args: any) => any) extends F ? F : any;
 
 declare function __VLS_normalizeSlot<S>(s: S): S extends () => infer R ? (props: {}) => R : S;
-declare function __VLS_componentProps<T, K>(comp: T, fnReturn: K):
-	__VLS_PickNotAny<K, {}> extends { __ctx: { props: infer P } } ? NonNullable<P>
-	: T extends (props: infer P, ...args: any) => any ? NonNullable<P> :
-	{};
-
-type __VLS_GetComponentProps<T> = NonNullable<NonNullable<T['__ctx']>['props']>;
-type __VLS_GetComponentPropsType<C, P> = __VLS_MappedOmit<Parameters<C>[0], (__VLS_IsAny<P> extends true ? never : keyof P) | keyof JSX.ElementChildrenAttribute> & __VLS_MappedOmit<__VLS_PickNotAny<P, {}>, keyof JSX.ElementChildrenAttribute> & Record<string, unknown>;
-declare function __VLS_cast<T>(x: T): T;
 `.trim();
 }
 

--- a/packages/vue-language-core/src/utils/globalTypes.ts
+++ b/packages/vue-language-core/src/utils/globalTypes.ts
@@ -118,6 +118,8 @@ declare function __VLS_componentProps<T, K>(comp: T, fnReturn: K):
 	__VLS_PickNotAny<K, {}> extends { __ctx: { props: infer P } } ? NonNullable<P>
 	: T extends (props: infer P, ...args: any) => any ? NonNullable<P> :
 	{};
+
+declare function __VLS_cast<T>(x: T): T;
 `.trim();
 }
 

--- a/packages/vue-language-core/src/utils/globalTypes.ts
+++ b/packages/vue-language-core/src/utils/globalTypes.ts
@@ -13,8 +13,6 @@ type __VLS_Element = __VLS_PickNotAny<import('vue/jsx-runtime').JSX.Element, JSX
 type __VLS_IsAny<T> = 0 extends 1 & T ? true : false;
 type __VLS_PickNotAny<A, B> = __VLS_IsAny<A> extends true ? B : A;
 
-type __VLS_MappedOmit<T, K> = { [P in keyof T as P extends K ? never : P]: T[P] };
-
 type __VLS_Prettify<T> = {
 	[K in keyof T]: T[K];
 } & {};

--- a/packages/vue-test-workspace/vue-tsc/non-strict-template/#3561/Comp.vue
+++ b/packages/vue-test-workspace/vue-tsc/non-strict-template/#3561/Comp.vue
@@ -6,7 +6,7 @@ defineProps<{
 }>();
 
 defineSlots<{
-	default: { item: T },
+	default(props: { item: string }): any;
 }>();
 //...
 </script>

--- a/packages/vue-test-workspace/vue-tsc/non-strict-template/#3561/Comp.vue
+++ b/packages/vue-test-workspace/vue-tsc/non-strict-template/#3561/Comp.vue
@@ -8,5 +8,4 @@ defineProps<{
 defineSlots<{
 	default(props: { item: string }): any;
 }>();
-//...
 </script>

--- a/packages/vue-test-workspace/vue-tsc/non-strict-template/#3561/Comp.vue
+++ b/packages/vue-test-workspace/vue-tsc/non-strict-template/#3561/Comp.vue
@@ -1,0 +1,12 @@
+<script setup lang="ts" generic="T, TValue extends undefined | string | number | symbol">
+defineProps<{
+	items: Array<T>,
+	modelValue: TValue,
+	getKey: (item: T) => TValue,
+}>();
+
+defineSlots<{
+	default: { item: T },
+}>();
+//...
+</script>

--- a/packages/vue-test-workspace/vue-tsc/non-strict-template/#3561/Comp.vue
+++ b/packages/vue-test-workspace/vue-tsc/non-strict-template/#3561/Comp.vue
@@ -6,6 +6,6 @@ defineProps<{
 }>();
 
 defineSlots<{
-	default(props: { item: string }): any;
+	default(props: { item: T }): any;
 }>();
 </script>

--- a/packages/vue-test-workspace/vue-tsc/non-strict-template/#3561/main.vue
+++ b/packages/vue-test-workspace/vue-tsc/non-strict-template/#3561/main.vue
@@ -15,7 +15,7 @@ const componentState = {
 		:items="componentState.items"
 		:get-key="(item) => {
 			exactType(item, null);
-			return ''
+			return '';
 		}"
 		class="mb-3"
 	>

--- a/packages/vue-test-workspace/vue-tsc/non-strict-template/#3561/main.vue
+++ b/packages/vue-test-workspace/vue-tsc/non-strict-template/#3561/main.vue
@@ -19,6 +19,6 @@ const componentState = {
 		}"
 		class="mb-3"
 	>
-		{{ exactType(item, null) }}
+		{{ exactType(item, '' as string) }}
 	</Comp>
 </template>

--- a/packages/vue-test-workspace/vue-tsc/non-strict-template/#3561/main.vue
+++ b/packages/vue-test-workspace/vue-tsc/non-strict-template/#3561/main.vue
@@ -19,6 +19,6 @@ const componentState = {
 		}"
 		class="mb-3"
 	>
-		{{ exactType(item, '' as string) }}
+		{{ exactType(item, null) }}
 	</Comp>
 </template>

--- a/packages/vue-test-workspace/vue-tsc/non-strict-template/#3561/main.vue
+++ b/packages/vue-test-workspace/vue-tsc/non-strict-template/#3561/main.vue
@@ -17,7 +17,6 @@ const componentState = {
 			exactType(item, null);
 			return '';
 		}"
-		class="mb-3"
 	>
 		{{ exactType(item, null) }}
 	</Comp>

--- a/packages/vue-test-workspace/vue-tsc/non-strict-template/#3561/main.vue
+++ b/packages/vue-test-workspace/vue-tsc/non-strict-template/#3561/main.vue
@@ -1,0 +1,24 @@
+<script setup lang="ts">
+import { exactType } from 'vue-tsc/shared';
+import Comp from './Comp.vue';
+
+const componentState = {
+	selected: '',
+	items: [null],
+};
+</script>
+
+<template>
+	<Comp
+		v-slot="{ item }"
+		v-model="componentState.selected"
+		:items="componentState.items"
+		:get-key="(item) => {
+			exactType(item, null);
+			return ''
+		}"
+		class="mb-3"
+	>
+		{{ exactType(item, null) }}
+	</Comp>
+</template>

